src/main/java/bio/terra/service/dataset/flight/ingest/IngestUtils.java

```diff
@@ -255,11 +263,71 @@ public static PdaoLoadStatistics getIngestStatistics(FlightContext context) {
         return false;
       };
 
-  public static void checkForLargeIngestRequests(int numLines, int maxIngestRows) {
+  public static Stream<JsonNode> getJsonNodesStreamFromFile(
+      GcsPdao gcsPdao,
+      ObjectMapper objectMapper,
+      IngestRequestModel ingestRequest,
+      Dataset dataset,
+      List<String> errors) {
+    return gcsPdao
+        .getGcsFilesLinesStream(
+            ingestRequest.getPath(), dataset.getProjectResource().getGoogleProjectId())
+        .map(
+            content -> {
+              try {
+                return objectMapper.readTree(content);
+              } catch (JsonProcessingException ex) {
+                errors.add(ex.getMessage());
+                return null;
+              }
+            })
+        .filter(Objects::nonNull);
+  }
+
+  public static long countBulkFileLoadModelsFromPath(
+      GcsPdao gcsPdao,
+      ObjectMapper objectMapper,
+      IngestRequestModel ingestRequest,
+      Dataset dataset,
+      List<String> fileRefColumnNames,
+      List<String> errors) {
+    try (var nodesStream =
+        IngestUtils.getBulkFileLoadModelsStream(
+            gcsPdao, objectMapper, ingestRequest, dataset, fileRefColumnNames, errors)) {
+      return nodesStream.count();
+    }
+  }
+
+  public static Stream<BulkLoadFileModel> getBulkFileLoadModelsStream(
+      GcsPdao gcsPdao,
+      ObjectMapper objectMapper,
+      IngestRequestModel ingestRequest,
+      Dataset dataset,
+      List<String> fileRefColumnNames,
+      List<String> errors) {
+    return IngestUtils.getJsonNodesStreamFromFile(
+            gcsPdao, objectMapper, ingestRequest, dataset, errors)
+        .flatMap(
+            node ->
+                fileRefColumnNames.stream()
+                    .map(
+                        columnName -> {
+                          JsonNode fileRefNode = node.get(columnName);
+                          if (fileRefNode != null && fileRefNode.isObject()) {
+                            return objectMapper.convertValue(fileRefNode, BulkLoadFileModel.class);
+                          } else {
+                            return null;
+                          }
+                        })
+                    .filter(Objects::nonNull));
```

Comment:
If it works, I think this might be better, and avoids introducing a `null` for filtering:
```suggestion
                    .map(node::get)
                    .filter(n -> n != null && n.isObject())
                    .map(n -> objectMapper.convertValue(n, BulkLoadFileModel.class))
```